Added new envs for Quantum Leap 0.8.0

modularTaco opened the pull request:
- removed `CRATE_DEFAULTDB` env var as it's not listed in the configuration options from Quantum Leap (see https://github.com/orchestracities/ngsi-timeseries-api/blob/master/docs/manuals/admin/configuration.md)
- added new env vars for Quantum Leap 0.8.0 from here https://github.com/orchestracities/ngsi-timeseries-api/blob/master/docs/manuals/admin/configuration.md
- added all configuration options from `values.yaml` to `Readme.md`
- Switch to make the chart usable for work queue deployments too. Basically all we needed was a K8s command switch which changes the command if the chart gets used for a work queue deployment.
- Quoting of some env var values---this avoids errors with old k8s and QL versions
- On demand Postgres init container. The base values files will always put in a Postgres init container. A chart user may want to override this, but removing the `init` stanza from their own value file would have no effect. This PR adds an `enable` flag to the `init` stanza so chart users can decide whether or not they want to have the init container.
- Docs to explain how to use this chart to do a standalone or work queue deployment.

chicco785 commented on charts/quantumleap/README.MD:
```diff
@@ -3,10 +3,12 @@
 #### Note: Quantum Leap version
+
 This chart is **not** fully compatible with Quantum Leap versions earlier
 than `0.7.6`. However, it is still possible to use this chart with an older
```
i think this needs to be changed to 0.8.0? did you test your chart change with 0.7.6?

modularTaco replied:
I've not tested it with 0.7.6 but i cannot see any reason why it won't work with 0.7.6.

As the only change I made are new envs (except for the removal of `CRATE_DEFAULTDB`, but this does not seem to be present in 0.7.6 either), a version of QL which does not know these envs will simply ignore them.

I've changed the default value for `config`, so for QL versions without the `QL_DEFAULT_DB` env you need to set the `config` variable to `{ tenants: {}, default-backend: "Crate" }`, as it was before, if you want to change the default backend.

Please correct me if I'm wrong

modularTaco replied:
I've made it clearly now in the readme, hope you agree

chicco785 commented on charts/quantumleap/README.MD:
```diff
@@ -3,10 +3,12 @@
-* Quantum Leap >= 0.7.6
+
+- Kubernetes 1.9+
+- Quantum Leap >= 0.7.6
```
se below

chicco785 commented:
thx @modularTaco for your contribution, i made two small related comments, have a look and let me know, then we can merge

chicco785 commented:
> Hey @chicco785, I've made the changes needed to run a QL version lower than 0.8.0 with this chart now clearly in the readme. Hope you agree

looks great! we only have one problem... travis ci has not yet reset the oss credits, so we cannot make the autobuild :/ are you in a hurry?

modularTaco commented:
> > Hey @chicco785, I've made the changes needed to run a QL version lower than 0.8.0 with this chart now clearly in the readme. Hope you agree
> 
> looks great! we only have one problem... travis ci has not yet reset the oss credits, so we cannot make the autobuild :/ are you in a hurry?

okay, no problem  
It's not urgent for me

modularTaco commented:
> > Hey @chicco785, I've made the changes needed to run a QL version lower than 0.8.0 with this chart now clearly in the readme. Hope you agree
> 
> looks great! we only have one problem... travis ci has not yet reset the oss credits, so we cannot make the autobuild :/ are you in a hurry?

Hey @chicco785 , any news here?

chicco785 commented:
@modularTaco so issue is that credits from travis are blocked. we had the same issue on https://github.com/orchestracities/ngsi-timeseries-api and moved to circle-ci, it was quite some pain and we don't have yet there the same feature level we had before, but so far we are happy. so we will move tests also here to circle-ci in the next days

modularTaco commented:
@chicco785 Updated the Fork to support new envs of QL 0.8.2 and use orchestracities repos instead of smartsdk

modularTaco commented:
@chicco785 any news here?

chicco785 commented:
> @chicco785 any news here?

#97 will fix the build mechanism, once merged, you should rebase if everything works fine, we will then merge ;)

c0c0n3 commented:
@modularTaco #97 just got merged, if you could please rebase on master, thanks alot!!

c0c0n3 reviewed (commented):
@modularTaco great work, thank you sooooooo much. I've only got some small requests for minor improvements if you could please have a look at the review comments and let me know if you'd like to do things differently instead.

modularTaco commented:
Hi @c0c0n3 , i've implemented your suggested changes (totally made sense to me) and rebased the branch. Let me know if you need anything from me for your tests. But please keep in mind that i'm not using QL anymore, so i won't be able to do any testing of the chart.

c0c0n3 commented:
hi @modularTaco, excellent job, you're a star. don't worry about testing, I'm actually busy setting up a test env on my side to be able to test the work queue deployment too, so i'll have to test the chart in the process anyway :-)

if you still have some dev cycles you could throw at this PR, it'd be a great help if you could fix up the linter issues---marked in the files changed tab. the linter is moaning about some whitespace and case stuff...

c0c0n3 commented:
@modularTaco oh my goodness, you're the fastest gun in the west! (talking about the linter issues)

c0c0n3 commented:
@modularTaco can you please rebase on master? I've merged your other PR for Crate. Thanks alot!!

c0c0n3 commented:
@modularTaco so I've implemented the changes I mentioned earlier and opened a PR in your fork

- https://github.com/modularTaco/charts/pull/1

if you could please merge that PR into your [ql080envs](https://github.com/modularTaco/charts/tree/ql080envs) branch so we can then go ahead w/ merging this PR (#90) into this repo: https://github.com/orchestracities/charts

Sorry it's a bit of a roundabout way to contribute my changes, but I couldn't find an easier way. But I guess all you need to do is merge my PR into your repo.

c0c0n3 commented:
### Testing

We don't have an infra for **automated** end-to-end testing at the moment, so I tested manually. I've managed to do both a standalone and work queue deployment using this chart---read the docs added to this PR for the details.

Here's a log of the test session for the work queue deployment---my env: K8s `1.16.15`, Helm `2.17.0`. First off, I whipped together two values files, one for the frontend and the other for the work queue backend:

```console
$ diff fe.values.yaml quantumleap/values.yaml
```
```diff
127c127
<     offloadWork: true
---
>     offloadWork: false
```
```console
$ diff wq.values.yaml quantumleap/values.yaml
```
```diff
132c132
<     workers: 2  # 0 means disabled
---
>     workers: 0  # 0 means disabled
```

Then I used the chart to make a new Helm release for the frontend deployment

```console
$ helm install quantumleap --name fe -f fe.values.yaml
NAME:   fe
LAST DEPLOYED: Thu Jul 21 15:12:33 2022
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/Deployment
NAME            READY  UP-TO-DATE  AVAILABLE  AGE
fe-quantumleap  0/2    0           0          0s

==> v1/Pod(related)
NAME                             READY  STATUS             RESTARTS  AGE
fe-quantumleap-65ff449d48-4x2cx  0/1    ContainerCreating  0         0s
fe-quantumleap-65ff449d48-spwjq  0/1    ContainerCreating  0         0s

==> v1/Service
NAME            TYPE       CLUSTER-IP  EXTERNAL-IP  PORT(S)  AGE
fe-quantumleap  ClusterIP  None        <none>       80/TCP   0s


NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace default -l "app=quantumleap,release=fe" -o jsonpath="{.items[0].metadata.name}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl port-forward $POD_NAME 8080:8668
```

Following that, I used the same chart again, with a different values file, to make a new Helm release for the work queue backend deployment:

```console
$ helm install quantumleap --name wq -f wq.values.yaml
NAME:   wq
LAST DEPLOYED: Thu Jul 21 15:15:20 2022
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/Deployment
NAME            READY  UP-TO-DATE  AVAILABLE  AGE
wq-quantumleap  0/2    0           0          0s

==> v1/Pod(related)
NAME                             READY  STATUS             RESTARTS  AGE
fe-quantumleap-65ff449d48-4x2cx  1/1    Running            0         2m46s
fe-quantumleap-65ff449d48-spwjq  1/1    Running            0         2m46s
wq-quantumleap-7996dd767f-ffjl8  0/1    ContainerCreating  0         0s
wq-quantumleap-7996dd767f-sr686  0/1    ContainerCreating  0         0s

==> v1/Service
NAME            TYPE       CLUSTER-IP  EXTERNAL-IP  PORT(S)  AGE
wq-quantumleap  ClusterIP  None        <none>       80/TCP   0s


NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace default -l "app=quantumleap,release=wq" -o jsonpath="{.items[0].metadata.name}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl port-forward $POD_NAME 8080:8668
```

At this point, as a sanity check I ran some smoke tests against all the QL pods (frontend and backend) in the K8s cluster and checked Helm was correctly tracking the two releases

```console
$ helm list
NAME	REVISION	UPDATED                 	STATUS  	CHART             	APP VERSION	NAMESPACE
fe  	1       	Thu Jul 21 15:12:33 2022	DEPLOYED	quantumleap-0.1.20	           	default  
wq  	1       	Thu Jul 21 15:15:20 2022	DEPLOYED	quantumleap-0.1.20	           	default  
```

Finally, I zapped both releases

```console
$ helm delete wq
release "wq" deleted

$ helm list
NAME	REVISION	UPDATED                 	STATUS  	CHART             	APP VERSION	NAMESPACE
fe  	1       	Thu Jul 21 15:12:33 2022	DEPLOYED	quantumleap-0.1.20	           	default  

$ helm delete fe
release "fe" deleted
```

c0c0n3 commented:
hi @modularTaco, just merged your PR, thanks again for the effort you put in, I appreciate that even more knowing you did this in your own time out of your kind heart :-)

modularTaco commented:
> hi @modularTaco, just merged your PR, thanks again for the effort you put in, I appreciate that even more knowing you did this in your own time out of your kind heart :-)

No problem. Better to invest free time instead of "loosing" the time needed to create these PRs last year 👍
Thanks for merging
